src/bitdrift_public/protobuf/workflow/v1/workflow.proto

```diff
@@ -133,6 +141,63 @@ message Workflow {
       ActionEmitMetric action_emit_metric = 2;
       ActionEmitSankeyDiagram action_emit_sankey_diagram = 3;
       ActionTakeScreenshot action_take_screenshot = 4;
+      ActionGenerateLog action_generate_log = 5;
+    }
+
+    // Generates a new log message. This log will be injected into the workflow engine and processed
+    // like any other log, either by this workflow or by another workflow.
+    message ActionGenerateLog {
+      // Describes how to obtain a value used in field composition.
+      message ValueReference {
+        oneof value_reference_type {
+          // Fixed value that never changes.
+          string fixed = 1;
+          // The value of the field from the current log.
+          string field_from_current_log = 2;
+          // The value of a field from a previous log, saved via the SaveField extension. This is
+          // the ID of the extension.
+          string saved_field_id = 3;
+          // The timestamp of a previous log, saved via the SaveTimestamp extension. This is the ID
+          // of the extension.
+          string saved_timestamp_id = 4;
+        }
+      }
+
+      // A pair of values.
+      message ValueReferencePair {
+        ValueReference lhs = 1;
+        ValueReference rhs = 2;
+      }
+
+      // A field to be generated. Note that when operating on timestamps, the values are expected
+      // to be fractions of seconds since the Unix epoch.
+      message GeneratedField {
+        // The name of the field.
+        string name = 1;
+
+        // The value of the field.
+        oneof generated_field_value_type {
+          option (validate.required) = true;
+
+          // A single value.
+          ValueReference single = 2;
+
+          // Perform subtraction. lhs - rhs. Both values must be convertible to floating point.
+          ValueReferencePair subtract = 3;
+
+          // Perform addition. lhs + rhs. Both values must be convertible to floating point.
+          ValueReferencePair add = 4;
+
+          // Perform multiplication. lhs * rhs. Both values must be convertible to floating point.
+          ValueReferencePair multiply = 5;
+
+          // Perform division. lhs / rhs. Both values must be convertible to floating point.
+          ValueReferencePair divide = 6;
+        }
+      }
+
+      string message = 1;
+      repeated GeneratedField fields = 2;
```

Comment:
what would be the log level and log type of the generated log? do you want that configurable?

Reply:
Yeah could question. We should probably make it configurable. @Reflejo ?

Reply:
probably let them set the log level but not the type? I think that's pretty opaque for user consumption

Reply:
Per offline convo with hard code to DEBUG for now we can revisit later if needed.